File: front-ui/src/services/orders/OrdersService.ts
import OrdersApi from '@api/orders/OrdersApi';
import { CreateOrderRequest, PaginatedOrdersResponse } from '@api/orders/OrderTypes';
import { HttpPromise } from 'simple-http-rest-client';

/**
 * OrdersService handles all business logic for market orders.
 * It provides methods for fetching orders and creating new ones.
 */
export default class OrdersService {
  constructor(private readonly ordersApi: OrdersApi) {}

  /**
   * Fetches orders with pagination support
   *
   * @param page The page number (1-based)
   * @param size The number of items per page
   * @returns A promise containing the paginated orders response
   */
  fetchOrders = (page: number, size: number): HttpPromise<PaginatedOrdersResponse> => this.ordersApi.fetchOrders(page, size);

  /**
   * Creates a new market order
   *
   * @param request The order details
   * @returns A promise containing the ID of the created order
   */
  createOrder(request: CreateOrderRequest): HttpPromise<number> {
    return this.ordersApi.createOrder(request);
  }

  /**
   * Deletes an order by its ID
   *
   * @param id The ID of the order to delete
   * @returns A promise that resolves when the order is deleted
   */
  deleteOrder(id: string): HttpPromise<void> {
    return this.ordersApi.deleteOrder(id);
  }
}

File: src/main/java/banky/services/orders/OrdersService.java
package banky.services.orders;

import banky.db.dao.OrdersDao;
import banky.db.generated.Orders;
import banky.webservices.api.orders.requests.CreateOrderRequest;
import banky.webservices.api.orders.responses.OrderResponse;
import banky.webservices.data.pagination.PaginatedResponse;
import banky.webservices.data.pagination.PaginationMeta;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;

import java.util.List;

import static banky.webservices.data.pagination.PaginationHelper.calculateTotalPages;

/**
 * Service responsible for order-related business logic.
 * Coordinates between web services and data access layer.
 */
@Singleton
public class OrdersService {
    private final OrdersDao ordersDao;

    @Inject
    private OrdersService(OrdersDao ordersDao) {
        this.ordersDao = ordersDao;
    }

    /**
     * Fetch orders with pagination support.
     * 
     * @param page The page number to retrieve (1-based)
     * @param size The number of items per page
     * @return A paginated response containing orders and pagination metadata
     */
    public PaginatedResponse<OrderResponse> fetchOrders(int page, int size) {
        // Calculate total elements and pages
        long totalElements = ordersDao.countOrders();
        int totalPages = calculateTotalPages(totalElements, size);
        
        // Ensure page is within bounds
        int adjustedPage = Math.min(Math.max(page, 1), Math.max(totalPages, 1));
        
        // Get orders for the page
        List<OrderResponse> orders = ordersDao.fetchOrders(adjustedPage, size);
        
        // Create pagination metadata
        PaginationMeta paginationMeta = new PaginationMeta(
            adjustedPage,
            totalPages,
            totalElements,
            size
        );
        
        return new PaginatedResponse<>(orders, paginationMeta);
    }

    /**
     * Creates a new order
     *
     * @param request The order details including date, name, amount, quantity, charges, accountId, and tickerId
     * @return The ID of the newly created order
     */
    public Long createOrder(CreateOrderRequest request) {
        Orders order = new Orders();
        
        order.setDate(request.date());
        order.setAmount(request.amount());
        order.setQuantity(request.quantity());
        order.setCharges(request.charges());
        order.setAccountId(request.accountId());
        order.setTickerId(request.tickerId());
        order.setSide(request.side().toString());

        return ordersDao.save(order).getId();
    }

    /**
     * Deletes an order by its ID
     *
     * @param orderId The ID of the order to delete
     */
    public void deleteOrder(Long orderId) {
        ordersDao.delete(orderId);
    }
}